### PR TITLE
TDD workflow document

### PR DESCRIPTION
**Pull request by ashetty1**

### CODE REVIEW

**Comment by kadel** [comment]
Wouldn't be development.md better place for this?
Or putting it into docs folder, and adding a few words to development.md with a link to this.

I'm afraid that it will be hidden in tests dir and no one will notice it.

@cdrage what do you think?

**Comment by ashetty1** [comment]
@kadel Oh yeah, putting this under docs certainly makes sense.

File: tests/TDD.md [hunk 1]
```diff
@@ -0,0 +1,91 @@
+## Test Driven Development (TDD) for ODO
+
+What’s TDD?  In simple terms, WRITE TESTS FIRST
```
Comment (cdrage) [inline tests/TDD.md]: remove spacing after `TDD?` there are double spaces there

File: tests/TDD.md [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+
+#### TDD Workflow
+
+Under TDD workflow, we write tests before writing feature code
+The tests written will serve as a blueprint for writing feature code
+
+1. Add a test
+1. Run all tests and see if the new test fails
```
Comment (cdrage) [inline tests/TDD.md]: although this is auto generated on GitHub, you need to change the numbers to 1,2,3,4,etc.

File: tests/TDD.md (hunk 1 above)
Comment (cdrage) [reply tests/TDD.md]: add a period end of sentence

File: tests/TDD.md [hunk 3]
```diff
@@ -0,0 +1,91 @@
+## Test Driven Development (TDD) for ODO
+
+What’s TDD?  In simple terms, WRITE TESTS FIRST
+
+#### TDD Workflow
+
+Under TDD workflow, we write tests before writing feature code
```
Comment (cdrage) [inline tests/TDD.md]: periods

File: tests/TDD.md [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+1. Run all tests and see if the new test fails
+1. Write the code
+1. Run tests
+1. Rewrite code incorporating any changes, and add new tests
+
+#### Purpose
+
+1. Writing tests before implementing the functionality will enable us to write tests covering more scenarios
+1. A feature with a good set of cases is also the most robust, thereby helping us improve the quality of the product
+1. Would help us address the test coverage backlog we have
```
Comment (cdrage) [inline tests/TDD.md]: don't really need this part. already covered the purpose in the first sentence of this document

File: tests/TDD.md [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+
+
+#### What kind of tests?
+
+1. Phase 1: Unit tests will be our primary focus. 
+1. Phase 2: Once we get comfortable with the process, we will work on incorporating e2e tests into this workflow.
+
```
Comment (cdrage) [inline tests/TDD.md]: No need to document phases, this part should be removed too

File: tests/TDD.md [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+	
+1. Review process
+	* Ensure that PR with code change is accompanied by unit tests
+	* Travis checks to be our point of determination
+
+#### What does following TDD process involve
+
+1. Reversing the process which we currently follow, i.e writing tests after writing the feature
+1. We should treat tests as something that will help us gauge end-user interaction and product behaviour
+
```
Comment (cdrage) [inline tests/TDD.md]: Still, got to try and keep the document short. Developers are the laziest people I know (that's a compliment!) and don't want to read a lot. Perhaps removing this part too would be helpful. No need to explain the history / why we need it.

File: tests/TDD.md [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+1. Would help us address the test coverage backlog we have
+
+
+#### What kind of tests?
+
+1. Phase 1: Unit tests will be our primary focus. 
+1. Phase 2: Once we get comfortable with the process, we will work on incorporating e2e tests into this workflow.
+
+
+#### TDD process for ODO
```
Comment (cdrage) [inline tests/TDD.md]: This part is good.

File: tests/TDD.md [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+
+1. Write tests for every feature/bug fix before implementing the code. 
+
+	* Developer will split the assigned task into small achievable subtasks
+	* Each subtask will require unit tests to be written before feature implementation
+	* Open work-in-progress [WIP] PR with the tests written
+	
+1. The tests should cover both positive and negative scenarios:
```
Comment (cdrage) [inline tests/TDD.md]: need to add numbers

File: tests/TDD.md [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+	* It would be good to have all permutations and combinations covered *as much as possible* helping us cover the boundary cases
+	* There is always scope to add  more tests could be added once the feature has been fully implemented
+	
+1. Write feature code and run tests:
+	* Feature code should be pushed to the same WIP test PR
```
Comment (cdrage) [inline tests/TDD.md]: what's "feature code"?

File: tests/TDD.md [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+
+1. Reversing the process which we currently follow, i.e writing tests after writing the feature
+1. We should treat tests as something that will help us gauge end-user interaction and product behaviour
+
+
+
+#### FAQs
+1. ##### Won't TDD slow down our development process?
+ 
+   Initially we might feel that way, but over time, we will realise the advantage this would bring to our development process by reducing the bugs, improving test coverage and the development time spent in fixing regression bugs.  TDD will help us treat test development as an independent process and an essential prerequisite in our development process.
```
Comment (cdrage) [inline tests/TDD.md]: `it might feel that way`

Comment (cdrage) [reply tests/TDD.md]: `spent fixing regression bugs`

Comment (cdrage) [reply tests/TDD.md]: Remove the last sentence, it seems like a fluff sentence / re-iterating what you already said.

File: tests/TDD.md [hunk 11]
```diff
@@ -0,0 +1,91 @@
...
+
+1. ##### What if we don’t know what a feature would finally look like
```
Comment (cdrage) [inline tests/TDD.md]: `What if we don't know what our code function will finally look like`

May be better? It's odd using `feature` everywhere. Or maybe that's just me.

File: tests/TDD.md [hunk 12]
```diff
@@ -0,0 +1,91 @@
...
+
+    Nothing prevents a developer from running the tests against his proof of concept code on his local setup before pushing the PR with tests.
```
Comment (cdrage) [inline tests/TDD.md]: huh, this doesn't really make sense. 

Maybe saying something along the lines of: "TDD is just a recommendation. Nothing is stopping you writing your code before your test. The objective of TDD is to teach developers to write tests *while* implementing your function".

Maybe I'm wrong haha.

File: tests/TDD.md [hunk 13]
```diff
@@ -0,0 +1,91 @@
...
+
+1. ##### Writing tests based on a feature is easier.
```
Comment (cdrage) [inline tests/TDD.md]: `based on a new feature` maybe?

File: tests/TDD.md [hunk 14]
```diff
@@ -0,0 +1,91 @@
...
+
+    Writing tests beforehand (under TDD) would help us identify the design limitations and corner cases in advance. Writing tests post development often limits us to cover positive cases. TDD will help us move away from that mindset.
+
+1. ##### Can we follow the same process for big features too?
+
+    Tasks involving big features should be split into numerous simple and achievable subtasks, and each of those subtasks should have test coverage. 
+
+1. ##### Are unit tests mandatory under TDD?
+  
+    Unit tests are key to TDD process. We will have unit tests (using fake client) as a mandatory requirement to PR review/acceptance.
```
Comment (cdrage) [inline tests/TDD.md]: Maybe remove this part.. @syamgk is contributing a doc with regards to how to write unit / integration tests. Mentioning `fake client` without any references will confuse the reader.

File: tests/TDD.md [hunk 15]
```diff
@@ -0,0 +1,91 @@
...
+
+1. ##### Should tests always be written first? 
+
+    That’s the point. In order to follow TDD process, we need to look at feature development from a testing perspective. It will help us cover wider scenarios beyond the positive scenarios. Tests will also serve as good validator during the coding process.
```
Comment (cdrage) [inline tests/TDD.md]: I think you meant to write `cover wider scenarios beyond positive cases`

File: tests/TDD.md [hunk 16]
```diff
@@ -0,0 +1,91 @@
...
+
+ 
+#### Learning resources
+1. https://github.com/quii/learn-go-with-tests
```
Comment (cdrage) [inline tests/TDD.md]: again, numbers.

also maybe add a small sentence here? `If you haven't tried TDD before, here are a few resources to get started!`

File: docs/tdd-workflow.md [hunk 17]
```diff
@@ -0,0 +1,65 @@
...
+
+Under TDD workflow, we write tests before writing feature code. Below is the workflow that is generally followed:
+
+1. Add a test
+2. Run all tests and see if the new test fails
+3. Write the code
+4. Run tests
+5. Rewrite code incorporating any changes, and add new tests
+
+
```
Comment (cdrage) [inline docs/tdd-workflow.md]: remove extra newline

File: docs/tdd-workflow.md [hunk 18]
```diff
@@ -0,0 +1,65 @@
...
+3. Write feature code and run tests:
+	* Feature code should be pushed to the same WIP test PR
+	* Feature code should pushed only after pushing the unit tests
+	* After the feature code is pushed, the test should pass
+	
+4. Review process
+	* Ensure that PR with code change is accompanied by unit tests
+	* Travis checks to be our point of determination
+
+
```
Comment (cdrage) [inline docs/tdd-workflow.md]: remove extra newline

File: docs/tdd-workflow.md [hunk 19]
```diff
@@ -0,0 +1,65 @@
...
+2. Run all tests and see if the new test fails
+3. Write the code
+4. Run tests
+5. Rewrite code incorporating any changes, and add new tests
+
+
+#### TDD process for ODO
+
+1. Write tests for every feature/bug fix before implementing the code. 
+
```
Comment (cdrage) [inline docs/tdd-workflow.md]: remove extra newline

File: docs/tdd-workflow.md [hunk 20]
```diff
@@ -0,0 +1,65 @@
...
+4. ##### Can we follow the same process for big features too?
+
+    Tasks involving big features should be split into numerous simple and achievable subtasks, and each of those subtasks should have test coverage. 
+
+5. ##### Should tests always be written first? 
+
+    That’s the point. In order to follow TDD process, we need to look at feature development from a testing perspective. It will help us cover wider scenarios beyond positive cases. Tests will also serve as good validator during the coding process.
+
+
```
Comment (cdrage) [inline docs/tdd-workflow.md]: remove xtra newline

**Comment by cdrage** [comment]
@kadel That works too :+1:

**Comment by ashetty1** [comment]
Think I have made all the changes, and have added a link to the development doc. @kadel @cdrage

**Comment by cdrage** [comment]
This doc LGTM!
